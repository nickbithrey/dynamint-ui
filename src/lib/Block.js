import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Block = ({children, ...remainingProps}) => (
	<Stack.Item {...remainingProps}>
		{children}
	</Stack.Item>
);

export default Block;