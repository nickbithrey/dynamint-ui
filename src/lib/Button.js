import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);