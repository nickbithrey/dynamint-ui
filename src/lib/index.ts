export interface Key {
    key: string | number;
}

export interface Condition<T> {
    condition: (input: T) => boolean;
}

export interface Loading {
    loading: boolean;
}

export interface Items<T> {
    items: Array<T>;
}

export interface URI {
    uri: string;
}

export interface Display {
    readOnly: boolean;
}

export interface IURILink {
    href: string;
}