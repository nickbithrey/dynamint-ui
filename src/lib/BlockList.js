import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'office-ui-fabric-react/lib/List';

const renderCell = (tag, updateEntry) => {
	const CellTag = tag;
	return (item, index) => {
		return <CellTag item={item} index={index} update={updateEntry} />;
	};
};

const BlockList = ({items, tag, updateEntry}) => {
	return (
		<List
	        items={items}
	        onRenderCell={renderCell(tag, updateEntry)}
	    />
	);
};

BlockList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object).isRequired,
	tag: PropTypes.func.isRequired
}

export default BlockList;