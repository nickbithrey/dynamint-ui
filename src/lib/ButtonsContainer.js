import React from 'react';
import Btn from '~/lib/Button';
import { Columnar, Block } from '~/lib/Grid';

const ButtonsContainer = ({children = [], buttons = []}) => (
	<Columnar horizontalAlign={'end'} tokens={{childrenGap: 5}}>
		{renderChildren(children)}
		{buttons.map(renderButton)}
	</Columnar>
);

const renderChildren = (children) => {
	if (Array.isArray(children)) {
		return children.map(renderChild)
	} else if (children) {
		return renderChild(children, 0);
	}
}

const renderChild = (child, i) => (
	<Block key={i}>
		{child}
	</Block>
);

const renderButton = (button, i) => {
	if (!button.condition()) {
		return;
	}
	return (
		<Block key={i}>
			<Btn {...button} />
		</Block>
	);
}

export default ButtonsContainer;