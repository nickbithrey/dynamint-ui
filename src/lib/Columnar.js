import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Columnar = ({children, ...remainingProps}) => (
	<Stack horizontal horizontalAlign="space-between" {...remainingProps}>
		{children}
	</Stack>
);

export default Columnar;