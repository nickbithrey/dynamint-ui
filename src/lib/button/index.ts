import { Button, IButton } from './Button';
import { ButtonContainer, IButtonContainer } from './ButtonContainer';

export { 
    Button,
    IButton,
    ButtonContainer,
    IButtonContainer
};