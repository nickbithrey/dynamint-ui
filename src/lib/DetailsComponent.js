import React from 'react';
import PropTypes from 'prop-types';
import { newBtn, newPrimaryBtn } from '~/lib/Button';
import Loading from '~/lib/Loading';
import { Rowise, Block } from '~/lib/Grid';
import ButtonsContainer from '~/lib/ButtonsContainer';

const detailsComponent = (WrappedComponent) => {
	class DetailsComponent extends React.Component {
		
		constructor(props) {
			super(props);
			this.load = this.load.bind(this);
			this.renderTitle = this.renderTitle.bind(this);
			this.createBackBtn = this.createBackBtn.bind(this);
			this.createSaveBtn = this.createSaveBtn.bind(this);
			this.back = this.back.bind(this);
			this.handleUpdateField = this.handleUpdateField.bind(this);
			this.handleUpdate = this.handleUpdate.bind(this);
			this.uri = props.location.state.uri;
			this.isCreate = props.location.state.isCreate;
		}
		
		load() {
			this.props.load(this.uri, this.isCreate);
		}
		
		componentDidMount() {
			this.load();
		}
		
		componentWillUnmount() {
			this.props.clear();
		}
		
		handleUpdateField(field) {
			this.props.updateField(field);
		}
		
		handleUpdate(e) {
			e.preventDefault();
			this.props.update(this.props.record, this.uri, this.isCreate)
			this.back();
		}
		
		createSaveBtn() {
			return newPrimaryBtn('Save', this.handleUpdate);
		}
		
		render() {
			if (this.props.loading || !this.props.record) {
				return <Loading />
			}
			const {
				updateField,
				update,
				...remainingProps
			} = this.props;
			return (
				<div>
					{this.renderTitle()}
					<Rowise tokens={{childrenGap: 5}}>
						<Block>
							<WrappedComponent fieldUpdate={this.handleUpdateField} update={this.handleUpdate} {...remainingProps} />
						</Block>
						<Block>
							<ButtonsContainer buttons={[this.createSaveBtn(), this.createBackBtn()]} />
						</Block>
					</Rowise>
				</div>
			);
		}
		
		renderTitle() {
			const {
				title
			} = this.props;
			return <h1>{title}</h1>;
		}
		
		back() {
			const {
				history
			} = this.props;
			history.goBack();
		}
		
		createBackBtn() {
			return newBtn('Back', this.back);
		}
		
	}

	DetailsComponent.propTypes = {
		load: PropTypes.func.isRequired,
		clear: PropTypes.func.isRequired,
		update: PropTypes.func.isRequired,
		updateField: PropTypes.func.isRequired,
		loading: PropTypes.bool.isRequired,
		history: PropTypes.object.isRequired,
		record: PropTypes.object,
		title: PropTypes.string.isRequired,
		url: PropTypes.string
	};
	
	return DetailsComponent;
}

export default detailsComponent;