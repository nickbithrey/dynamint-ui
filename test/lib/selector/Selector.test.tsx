import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ISelection, SelectionMode, Selection } from 'office-ui-fabric-react/lib/Selection';

import {
    ISelector,
    Selector,
    ISelectionButton,
    ISelectionDetails,
    SelectionType,
    ISelectorState
} from 'lib/selector/Selector';

const SubComp = () => <p>text</p>;

describe('selector', () => {
    
    const props: ISelector = {
        storeKey: 'store-key',
        loading: false,
        items: [
            {key: 'i1', selected: false},
            {key: 'i2', selected: false},
            {key: 'i3', selected: false}
        ],
        columns: [
            {name: 'col1', key: 'col1', fieldName: 'col1', minWidth: 50},
            {name: 'col2', key: 'col2', fieldName: 'col2', minWidth: 100}
        ],
        buttons: [
            {
                button: {
                    key: 'standard',
                    content: 'Standard', 
                    type: 'primary', 
                    onClick: jest.fn(), 
                    link: 'link'
                },
                condition: () => true
            },
            {
                button: {
                    key: 'standard',
                    content: 'Link Resolver', 
                    type: 'primary', 
                    onClick: jest.fn(), 
                    link: 'link'
                },
                linkResolver: (selectionDetails: ISelectionDetails) => 'updatedLink',
                condition: () => true
            },
            {
                button: {
                    key: 'standard',
                    content: 'onClick Resolver', 
                    type: 'primary', 
                    onClick: jest.fn(), 
                    link: 'link'
                },
                onClickResolver: (selectionDetails: ISelectionDetails) => jest.fn(),
                condition: () => true
            }
        ],
        selectionType: SelectionType.single
    };

    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = shallow(<Selector {...props} />);
    });
    
    it('will render a valid tag and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    it('initialised the state correctly', () => {
        expect(wrapper.state()).toEqual({
            selectionDetails: {
                selectedCount: 0,
                selected: [],
                selectedIndices: []
            }
        });
    });
    
    it('updates the state on selection change', () => {
        const indexToSelect = 0;
        const selection: ISelection = wrapper.find('Table').prop('selection');
        selection.setItems(props.items, true);

        selection.toggleIndexSelected(indexToSelect);

        expect(wrapper.instance().state).toEqual({
            selectionDetails: {
                selectedCount: 1,
                selected: [props.items[indexToSelect]],
                selectedIndices: [indexToSelect]
            }
        });
        const key = 'dynamint-ui-selector-' + props.storeKey;
        expect(JSON.parse(window.localStorage.getItem(key))).toEqual({
            selectionDetails: {
                selectedCount: 1,
                selected: [props.items[indexToSelect]],
                selectedIndices: [indexToSelect]
            }
        });
    });
    
    it('will clear the window storage when unmounting', () => {
        const key = 'dynamint-ui-selector-' + props.storeKey;
        window.localStorage.setItem(key, 'store');
        expect(window.localStorage.getItem(key)).toEqual('store');
        
        wrapper.unmount();
        
        expect(window.localStorage.getItem(key)).toBeNull();
    });
    
    it('will store the state to storage on update', () => {
        const key = 'dynamint-ui-selector-' + props.storeKey;
        window.localStorage.removeItem(key);
        expect(window.localStorage.getItem(key)).toBeNull();
        wrapper.instance().componentDidUpdate({}, {});
        expect(window.localStorage.getItem(key)).not.toBeNull();
    });

});
