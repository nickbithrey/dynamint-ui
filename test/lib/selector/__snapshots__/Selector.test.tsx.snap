// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selector will render a valid tag and matches snapshot 1`] = `
<Fragment>
  <Table
    columns={
      Array [
        Object {
          "fieldName": "col1",
          "key": "col1",
          "minWidth": 50,
          "name": "col1",
        },
        Object {
          "fieldName": "col2",
          "key": "col2",
          "minWidth": 100,
          "name": "col2",
        },
      ]
    }
    items={
      Array [
        Object {
          "key": "i1",
          "selected": false,
        },
        Object {
          "key": "i2",
          "selected": false,
        },
        Object {
          "key": "i3",
          "selected": false,
        },
      ]
    }
    loading={false}
    selection={
      Selection {
        "_anchoredIndex": 0,
        "_canSelectItem": [Function],
        "_changeEventSuppressionCount": 0,
        "_exemptedCount": 0,
        "_exemptedIndices": Object {},
        "_getKey": [Function],
        "_isModal": false,
        "_items": Array [],
        "_keyToIndexMap": Object {},
        "_onSelectionChanged": [Function],
        "_selectedItems": null,
        "_unselectableCount": 0,
        "_unselectableIndices": Object {},
        "count": 0,
        "mode": 1,
      }
    }
  />
  <Component
    buttons={
      Array [
        Object {
          "content": "Standard",
          "key": "standard",
          "link": "link",
          "onClick": [MockFunction],
          "type": "primary",
        },
        Object {
          "content": "Link Resolver",
          "key": "standard",
          "link": "updatedLink",
          "onClick": [MockFunction],
          "type": "primary",
        },
        Object {
          "content": "onClick Resolver",
          "key": "standard",
          "link": "link",
          "onClick": [MockFunction],
          "type": "primary",
        },
      ]
    }
  />
</Fragment>
`;
